pkg/kubectl/util/hash/hash.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hash
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"sort"
+
+	"k8s.io/kubernetes/pkg/api"
+)
+
+// ConfigMapHash returns a hash of the EncodeConfigMap encoding of `cm`
+func ConfigMapHash(cm *api.ConfigMap) string {
+	return encodeHash(hash(encodeConfigMap(cm)))
+}
+
+// SecretHash returns a hash of the EncodeSecret encoding of `sec`
+func SecretHash(sec *api.Secret) string {
+	return encodeHash(hash(encodeSecret(sec)))
+}
+
+// encodeConfigMap encodes a ConfigMap by encoding the Data with EncodeMapStringString and wrapping with {} braces
+func encodeConfigMap(cm *api.ConfigMap) string {
+	return fmt.Sprintf("{data:%s,}", encodeMapStringString(cm.Data))
+}
+
+// encodeSecret encodes a Secret by encoding the Data with EncodeMapStringByte and
+// appending ",type:"+sec.Type+"," to the encoded string, and wraps with {} braces
+func encodeSecret(sec *api.Secret) string {
+	s := encodeMapStringBytes(sec.Data)
+	return fmt.Sprintf("{data:%s,type:%s,}", s, string(sec.Type))
+}
+
+// encodeMapStringString extracts the key-value pairs from `m`, sorts them in byte-alphabetic order by key,
+// and encodes them in a string representation. Keys and values are separated with `:` and pairs are separated
+// with `,`. If m is non-empty, there is a trailing comma in the pre-hash serialization. If m is empty,
+// there is no trailing comma. The entire encoding starts with "{" and ends with "}".
+func encodeMapStringString(m map[string]string) string {
+	kv := make([][]string, len(m))
```

Comment:
`[]string` should be enough? Store keys of `m` in it and sort those keys. You can get value from `m[key]` when encoding to a string